### PR TITLE
Reduce memory usage of debug_trace* calls #9584

### PR DESCRIPTION
**Pull request by AliZDev-v0**
## PR description
Before on every opcode execution the memory was captured:
image

After if it's not memory writing opcode referring to previous captured memory:
image

## Fixed Issue(s)
Fixes Reduce memory usage of debug_trace* calls #9584

### Thanks for sending a pull request! Have you done the following?

- [x] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [x] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [x] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [x] spotless: `./gradlew spotlessApply`
- [x] unit tests: `./gradlew build`
- [x] acceptance tests: `./gradlew acceptanceTest`
- [x] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`
- [ ] hive tests: [Engine or other RPCs modified?](https://lf-hyperledger.atlassian.net/wiki/spaces/BESU/pages/22156302/Using+Hive+Test+Suite)

### CODE REVIEW

**Comment by macfarla** [comment]
@AliZDev-v0 some comments from a quick review:
  Test quality
  - Two good tests: one for the reuse path (isSameAs check) and one verifying a new snapshot is taken after a memory write
  - Tests use isSameAs (reference equality) which is the right assertion for this optimization
  - Missing: no test for the edge case where lastFrame is null (first opcode writes to non-empty memory) — though the guard handles this, a test would be nice
  - Missing: no test for cross-depth behavior — when a CALL returns and depth decreases, lastFrame is from the inner call. Is the previous frame's memory correct for the outer frame? The lastFrame is set in tracePostExecution which runs for every
  opcode regardless of depth, so after a CALL returns, lastFrame would be from the inner frame's final opcode — which has different memory than the outer frame. If the outer frame's next opcode is non-memory-writing, it would incorrectly reuse the
  inner frame's memory.

  Issues

  1. Cross-depth memory reuse bug: When execution returns from a nested CALL, lastFrame points to the inner call's last trace frame. If the next opcode in the outer frame doesn't write memory, captureMemory would return the inner frame's memory
  snapshot — which is wrong (different memory space). This needs a depth check, e.g.:
  } else if (frame.getMaybeUpdatedMemory().isEmpty()
      && lastFrame != null
      && lastFrame.getDepth() == frame.getDepth()) {
  2. Memory expansion without write: As noted above, implicit memory expansion (from gas cost calculations) won't trigger a new snapshot. This may cause trace output to show fewer memory words than expected. Lower severity since the "missing" words
  are all zeros.

  Suggestions

  - Add a depth check to the reuse condition to prevent cross-frame-depth memory contamination
  - Add a memory size check: && lastFrame.getMemory().map(m -> m.length).orElse(0) == frame.memoryWordSize() as a safety net
  - Add a test for the CALL/RETURN scenario

**Comment by AliZDev-v0** [comment]
@macfarla Thank you for review,
* Added a depth check 
* Added a memory size check:
* Add a test for the CALL/RETURN scenario (_shouldHandleCallScenarioAndDepthChange_)

**Comment by macfarla** [comment]
looks pretty ok to me, @ahamlat can you take a look since you raised the issue that this is addressing

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/DebugOperationTracer.java [hunk 1]
```diff
@@ -286,6 +286,11 @@ private Optional<Map<UInt256, UInt256>> captureStorage(final MessageFrame frame)
   private Optional<Bytes[]> captureMemory(final MessageFrame frame) {
     if (!options.traceMemory() || frame.memoryWordSize() == 0) {
       return Optional.empty();
+    } else if (frame.getMaybeUpdatedMemory().isEmpty()
+        && lastFrame != null
+        && lastFrame.getDepth() == frame.getDepth()
+        && lastFrame.getMemory().map(m -> m.length).orElse(0) == frame.memoryWordSize()) {
```
Comment (ahamlat) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/DebugOperationTracer.java]: This check has some overhead on the hot path, `.map(m -> m.length) `creates a lambda, boxes the int to Integer (auto-boxing), allocates a new Optional<Integer>, then .orElse(0) unboxes it back.

I would suggest something like this :
```
else if (lastFrame != null
    && lastFrame.getDepth() == frame.getDepth()
    && frame.getMaybeUpdatedMemory().isEmpty()
    && lastFrame.getMemory().isPresent()
    && lastFrame.getMemory().get().length == frame.memoryWordSize()) {
```

Comment (AliZDev-v0) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/DebugOperationTracer.java]: fixed, thank you
